src/webgpu/shader/execution/expression/binary/bitwise.spec.ts

```diff
@@ -3,82 +3,191 @@ Execution Tests for the bitwise binary expression operations
 `;
 
 import { makeTestGroup } from '../../../../../common/framework/test_group.js';
+import { assert } from '../../../../../common/util/util.js';
 import { GPUTest } from '../../../../gpu_test.js';
-import { i32, scalarType, u32 } from '../../../../util/conversion.js';
-import { allInputSources, run } from '../expression.js';
+import { abstractInt, i32, Scalar, scalarType, u32 } from '../../../../util/conversion.js';
+import { allInputSources, onlyConstInputSource, run } from '../expression.js';
 
-import { binary, compoundBinary } from './binary.js';
+import { abstractIntBinary, binary, compoundBinary } from './binary.js';
 
 export const g = makeTestGroup(GPUTest);
 
-function makeBitwiseOrCases(inputType: string) {
-  const V = inputType === 'i32' ? i32 : u32;
-  const cases = [
-    // Static patterns
+/**
+ * Collection of functions and values required to implement bitwise tests for a
+ * specific scalar type
+ */
+interface ScalarImpl {
+  // builder is a mostly a wrapper around type builders like 'i32' that handles
+  // the (number|bigint) type check.
+  builder: (value: bigint | number) => Scalar;
+  size: 32 | 64;
+}
+
+const kScalarImpls = {
+  i32: {
+    builder: (value: bigint | number): Scalar => {
+      assert(typeof value === 'number');
+      return i32(value);
```

Comment:
Is this right? [`i32()`](https://github.com/zoddicus/cts/blob/7ad3a97e893762d58f37499da327ef1ada6a503a/src/webgpu/util/conversion.ts#L1097) takes `value` and assigns it to [`workingDataI32`](https://github.com/zoddicus/cts/blob/7ad3a97e893762d58f37499da327ef1ada6a503a/src/webgpu/util/conversion.ts#L84). 
As all numbers in JS are float64, I'd expect `0b11111111111111111111111111111111` to be `2147483647` for both the `i32` and `u32` types. Assigning `2147483647`, which is larger than the (largest representable signed int32) to an element of `Int32Array` seems iffy. I can't find JS spec to know if this is safe or not.

For these bit patterns, I think it would be preferable to convert u32 bits to i32.

Reply:
I have uploaded a diff that changes these to using fooBits() instead of  foo()